source/functions.rst

```diff
@@ -1,191 +1,186 @@
-
-We will use pythontutor to exercise visualising program execution.
-
-|py-function-ns|
-
-.. |py-function-ns| raw:: html
-
-    <iframe width="800" height="500" frameborder="0" src="http://pythontutor.com/iframe-embed.html#code=x+%3D+1%0Ay+%3D+2%0Asuccess+%3D+'works'%0Afailure+%3D+'broken'%0A%0Adef+inc(p)%3A%0A++++incremented+%3D+p+%2B+1%0A++++return+incremented%0A%0Adef+print_result(result)%3A%0A++++if+result%3A%0A++++++++print(success)%0A++++else%3A%0A++++++++print(failure)%0A%0Ainc_x+%3D+inc(x)%0Aprint_result(inc_x+%3D%3D+y)%0A&origin=opt-frontend.js&cumulative=false&heapPrimitives=false&drawParentPointers=false&textReferences=false&showOnlyOutputs=false&py=2&rawInputLstJSON=%5B%5D&curInstr=0&codeDivWidth=350&codeDivHeight=400"> </iframe>
-
-Step through each line of code in the browser.
-
-Notice that when execution enters a function, a new 'frame' is
-created.
-
-The interpreter creates a new namespace associated with this frame. It is
-isolated from the 'parent' frame's namespace. This namespace is emtpy unless 
-parameters are passed.
+    def draw_polygon(side_length, sides):
+        for _ in range(sides):
+            turtle.forward(side_length)
+	    turtle.left(360.0/sides)
 
 .. tip::
 
-    A namespace and a frame are different objects. For the purpose of this course 
-    however think of them as the same.
+   See what happens if you define a function with and argument which has the
+   same name as a variable in your program.
 
+       >>> name = "Miss Moneypenny"
+       >>> def witty_comeback(name):
+       ...     print("Do you expect me to talk?")
+       ...     print("No " + name + ", I expect you to die!")
+       ...
+       >>> witty_comeback("Mr Bond")
+       Do you expect me to talk?
+       No ??? I expect you to die!
 
 Exercises
 =========
 
+Shapes
+------
 
-Shapes with Arguments
----------------------
-
-Reopen ``shapes.py`` and define every shape as a function with sensible arguments.
-
-Consider whether this make the code more modular, readable, reusable and general?
+1. Open your "shapes.py" file and define every shape as a function with
+   arguments.
+2. Write a simple program which uses your new functions to check that they work.
+3. Rewrite your program for drawing a house using your shape functions.
 
-House
------
 
-Refactor (rewrite) your house code as a function that uses two other functions.
+Conversions
+-----------
 
+4. Write a function called ``celsius_to_fahrenheit`` which has one argument and
+   convert a temperature in degrees celsius into the equivalent in degrees
+   fahrenheit
+   
+   .. tip:: Check Wikipedia to find out how to do this conversion --
+            https://en.wikipedia.org/wiki/Fahrenheit#Definition_and_conversions
 
-Conversion Programs
--------------------
+5. (Extension) Create a new program called "currency_converter.py" and copy the
+   following code into it::
 
-This exercise assumes you have completed the `about_functions` koans.
+     import urllib2 import json
 
```

Comment:
If import hasn't been covered before perhaps a link to something on import in python for those who are interested?